### ISSUE TITLE
Add instruction to forcefully refresh state from the script

### ISSUE THREAD

**Issue by MarcelT-NL** [hashicorp/terraform#23388]
### Current Terraform Version
0.12.14

### Use-cases
It would be great if you could issue a forced “refresh” at the top of the script, forcing the saved state to be deleted and recreated from the live environment.

Rationale:

Sometimes a terraform script fails, for whatever reason and you need to delete some assets manually (eg. in the Azure portal). When using Terraform.io (Cloud), there is no way to run CLI commands so no refresh option. Sometimes you have to delete and recreate the workspace because the state is out of sync with reality and terraform consistently fails when running.

Being able to have the state forcefully refreshed would solve a lot of manual work.

### Attempted Solutions
-

### Proposal
Add an instruction that can be used in scripts, that initiates a forced refresh of the state:

terraform_state_refresh

### References
-

**Comment by teamterraform** [comment]
Hi @MarcelT-NL! Thanks for reporting this.

We'd like to understand a bit more about the problem you've encountered here. Terraform automatically runs the same behavior as `terraform refresh` would take as an early part of `terraform plan`, so we're not sure what exactly it would mean to force a refresh from the configuration; it sounds like you've run into a specific problem that a normal `terraform refresh` wouldn't have resolved anyway, because otherwise `terraform plan` would automatically resolve it too.

It would help if you could share some full error messages you've seen in situations like what you described where a failure on a previous Terraform run caused you to make changes outside of Terraform that then caused Terraform to get "stuck". That would help us to determine what is failing in the refresh step that is preventing Terraform from recovering automatically.

Thanks!

**Comment by MarcelT-NL** [comment]
Hi,

It happens on different occasions. For instance, when you tear down an infrastructure with terraform, but some resource could not be deleted. Eg. a load balancer failed to be deleted because something was still connected to it. Seems to be related to race conditions when deleting infrastructure. Sometimes it happens when you rename a resource and terraform plan cannot remove (find) the old one. The only solution is to manually delete the resource. But then the state is out of sync with reality and fails consistently.

When I encounter this bug again I will add the error messages, even though I remember that they are generic.

**Comment by MarcelT-NL** [comment]
@teamterraform Did you get a chance to review my reply?

**Comment by danieldreier** [comment]
@MarcelT-NL most of the team will be out for the rest of this week for the US Thanksgiving holiday. I've put this on our internal list to follow up when we're back.

**Comment by danieldreier** [comment]
@MarcelT-NL If I understand your description correctly, the root issue you're trying to fix here is that you've encountered a bug that causes resource destruction to fail. Am I understanding you right? 

If so, my impulse is to focus on reproducing and fixing that, rather than doing work to make it easier to recover from the broken state you find yourself in.

**Comment by MarcelT-NL** [comment]
Yes you are correct. I tried reproducing the issue but have not encountered it in the past week.

One use case is:
- create some resources
- rename are least to resources
- plan and apply again
- the state contains the old name, the resource is not changed or destroyed from terraform
- after manually destroying the resource in azure: terraform cannot find the resource with the old name, fails.

This happened when creating and changing load balancers and some other instances. It may be that the names of 2 resources changed, when there was a dependency between them.
